### PR TITLE
Add PDOStatement with default fetch mode. Resolves #8374

### LINKED ISSUE #8374: PDOStatement with default fetch mode
* [ ] bug
* [x] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 3.2.3
* Platform and Target: not relevant

### What you did
Getting results out of a query
```php
$results = $query->execute()->fetchAll('assoc');
```
### Expected Behavior

I need 'assoc' as default. 
Passing 'assoc' all the time is redundant code and hard to change in the future.
I want to set a default fetch style for all statements (like in PDO).
http://php.net/manual/en/pdostatement.setfetchmode.php

I expect that fetchAll() and fetch() returns my custom 'default' fetch style.

```php
$results = $query->execute()->fetchAll();
```
### Actual Behavior
At the moment fetchAll and fetch() has PDO::FETCH_NUM as default fetch mode.
https://github.com/cakephp/database/blob/master/Statement/PDOStatement.php

```php
public function fetch($type = 'num')
```
```php
public function fetchAll($type = 'num')
```

### PR DESCRIPTION
**Pull request by odan**
- Added function setFetchMode to set the default fetch mode for a statement.
- Added function getFetchMode to get fetch mode for a statement.
- Added new connection option "fetchMode" to set default fetch mode for all select queries.

### CODE REVIEW

File: src/Database/Connection.php [hunk 1]
```diff
@@ -694,38 +699,6 @@ public function log($sql)
     }
 
     /**
-     * Check if cross talk is supported between two connections
-     *
-     * @param ConnectionInterface $target Connection to check cross talk with
-     *
-     * @return bool
-     */
-    public function supportsCrossWith(ConnectionInterface $target)
```
Comment (lorenzo) [inline src/Database/Connection.php]: I assume this was not intentional

Comment (odan) [inline src/Database/Connection.php]: This was not my intention. I'll fix that.

File: src/Database/Statement/StatementDecorator.php [hunk 2]
```diff
@@ -319,4 +326,26 @@ public function getInnerStatement()
     {
         return $this->_statement;
     }
+
+    /**
+     * Set the default fetch mode for this statement
+     *
+     * @param string $mode The fetch mode as string 'num' or 'assoc'
+     * @return bool true on success, false otherwise
```
Comment (antograssiot) [inline src/Database/Statement/StatementDecorator.php]: there's no return as stated by [travis](https://travis-ci.org/cakephp/cakephp/jobs/114062082#L242-L248)

**Comment by antograssiot** [comment]
usually the setter and getter is combined. Can't we apply the same logic here instead of setFetchMode et getFetchMode ?

File: src/Database/Statement/PDOStatement.php [hunk 3]
```diff
@@ -114,8 +115,9 @@ public function fetch($type = 'num')
      * @param string $type num for fetching columns as positional keys or assoc for column names as keys
      * @return array list of all results from database for this statement
      */
-    public function fetchAll($type = 'num')
+    public function fetchAll($type = null)
     {
+        $type = $this->getFetchMode($type);
```
Comment (markstory) [inline src/Database/Statement/PDOStatement.php]: Could these be `$type = $type ?: $this->getFetchMode();` This makes the getFetchMode() method a simple getter, and avoids additional function calls for the primary use cases in the ORM.

**Comment by lorenzo** [comment]
ping @odan

File: src/Database/Statement/PDOStatement.php (hunk 3 above)
Comment (odan) [inline src/Database/Statement/PDOStatement.php]: Yes good idea! I can fix it.

**Comment by odan** [comment]
I try to fix the CI errors.

**Comment by markstory** [comment]
Closing as this can't be merged anymore. If you're still interested in getting it merged, please rebase and open a new pull request, or re-open this one

**Pull request closed without merging**